arbiter/drivers/onedrive.cpp

```diff
@@ -0,0 +1,292 @@
+
+                //add result of children processes to the parent
+                result.insert(result.end(), children.begin(), children.end());
+            }
+        }
+
+        //check for link to next set
+        if (obj.contains("@odata.nextLink"))
+            pageUrl = obj.at("@odata.nextLink");
+        else
+            break;
+
+    } while (true);
+
+    return result;
+}
+
+std::vector<std::string> OneDrive::glob(std::string path, bool verbose) const
+{
+    path.pop_back();
+    bool recursive(path.back() == '*');
+    if (recursive)
+        path.pop_back();
+
+    if (path.back() == '/')
+        path.pop_back();
+
+    return processList(path, recursive);
+}
+
+bool OneDrive::get(
+        const std::string path,
+        std::vector<char>& data,
+        const http::Headers userHeaders,
+        const http::Query query) const
+{
+    Resource resource(path);
+    http::Headers headers(m_auth->headers());
+    headers["Content-Type"] = "application/octet-stream";
+    headers.insert(userHeaders.begin(), userHeaders.end());
+
+    drivers::Https https(m_pool);
+    const auto res(https.internalGet(resource.binaryEndpoint(), headers));
+
+    if (!res.ok()) {
+        std::cout <<
+            "Failed get - " << res.code() << ": " << res.str() << std::endl;
+        return false;
+    }
+
+    data = res.data();
+    return true;
+}
+
+OneDrive::Auth::Auth(std::string s) {
+    const auto config = json::parse(s);
+    m_token = config.at("access_token").get<std::string>();
+    m_refresh = config.at("refresh_token").get<std::string>();
+    m_redirect = config.at("redirect_uri").get<std::string>();
+    m_id = config.at("client_id").get<std::string>();
+    m_secret = config.at("client_secret").get<std::string>();
+    m_expiration = Time().asUnix();
+}
+
+std::unique_ptr<OneDrive::Auth> OneDrive::Auth::create(const std::string s)
+{
+    return makeUnique<Auth>(s);
+}
+
+void OneDrive::Auth::refresh()
+{
+    //only refresh if we get within 2 minutes of the token ending
+    const auto now(Time().asUnix());
+    if (m_expiration - now > 120)
+        return;
+
+    http::Pool pool;
+    drivers::Https https(pool);
+    http::Headers headers({ });
+
+    headers["Accept"] = "application/json";
+    headers["Content-Type"] = "application/x-www-form-urlencoded";
+
+    json bodyJson = {
+        { "access_token", m_token },
+        { "refresh_token", m_refresh },
+        { "client_id", m_id },
+        { "client_secret", m_secret },
+        { "scope", "offline_access+files.read.all+user.read.all" },
+        { "grant_type", "refresh_token" }
+    };
+
+    //create url-encoded body
+    std::string body = "";
+    for (auto& i: bodyJson.items())
+        body += i.key() + "=" + i.value().get<std::string>() + "&";
+    body.pop_back();
+    std::vector<char> content(body.begin(), body.end());
+
+    const auto res(https.internalPost(this->getRefreshUrl(), content, headers));
```

Comment:
Actually, why are we passing this as form-encoded body data?  It would be more clear if you make your `bodyJson` a `Query` (which is just a string map) and then pass it to `internalPost` as the query parameters instead of the query-stringified body.  This would make the `buildQueryString` comment above moot.

Reply:
This part has to be passed in the body as a url-encoded string that looks exactly like a query. Tried again just now, just in case, and confirmed that.

Reply:
`The request body must contain the following parameter: 'grant_type'`

Reply:
You removed the Content-Type of `application/x-www-form-urlencoded` for that test?

Reply:
Refer to the other finding in that case.